File: symmetric-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
# 
# call(r.l, r.r): (only if root is not nil)
#   func(l, r)
#     if both nil return true
#     if one is nil return false
#     if l.val != r.val return false
#     return both of: 
#         call(l.l, r.r) 
#         call(l.r, r.l) 
#      1
#    2   2
class Solution:
    def sym(self, left: Optional[TreeNode], right: Optional[TreeNode]) -> bool:
        if not left and not right:
            return True
        if not left or not right:
            return False
        if left.val != right.val:
            return False
        return self.sym(left.left, right.right) and self.sym(left.right, right.left)

    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        return self.sym(root.left, root.right)
        

File: lc/go/symmetric-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
//  left and right
// bfs where the queue is equal to its reverse for each row
// dfs: ...
// preorder equals postorder is the guaranteed way to solve this -- fail
//     1
//    2   x
// q = [2, x]
// ch = [2, x]
func reverses(a []*TreeNode) bool {
    n:=len(a)
    for i,j:=0,n-1; i<j; i,j=i+1,j-1 {
        aival, ajval := 101, 101
        if a[i] != nil { aival = a[i].Val }
        if a[j] != nil { ajval = a[j].Val }
        if aival != ajval { return false }
    }
    return true
}
func isSymmetric(root *TreeNode) bool {
    q := []*TreeNode{root}
    for ; len(q) > 0; {
        if !reverses(q) { return false }
        ch := []*TreeNode{}
        for _,n := range q {
            if n == nil { continue } 
            ch = append(ch, n.Left)
            ch = append(ch, n.Right)
        }
        q = ch
    }
    return true
}